Build CBFlib from fork

ndevenish opened the pull request:
This builds cbflib, and my manual feedstock (mac, linux) packages can be tested with `conda install -c ndevenish cbflib`. It's currently using a heavily-reworked [CMakeLists.txt](https://github.com/ndevenish/cbflib/blob/refactor/CMakeLists.txt) on a custom branch (the original didn't seem to work and looked out of date), but this can be moved in-recipe if necessary.

The `CMakeLists.txt` primary changes make it more like a usual library project - it doesn't build it's own dependencies, only build static _or_ shared libs per [`BUILD_SHARED_LIBS`](https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html) (making it easier to conform with [CFEP-18](https://github.com/conda-forge/cfep/blob/master/cfep-18.md)) and has toggles for Fortran, TIFF dependency, pycbf building and a few others. It builds `_pycbf.so` itself and uses (a new) `setup.py` to just install the files, contained in a subpackage rather than all in `site_packages/*`. This is done so that the entire build environment doesn't need to be carefully managed for python to build it itself. For conda this works fine, but might need revisiting externally. There's also a few parts of the behaviour I don't understand - like the LZ4 and BSHUF handling. Maybe this is just for manual addition of the filters for static linking.

There are a few caveats to the build here:
- The `pycbf_wrap.c` and `pycbf.py` files need to be regenerated with SWIG 4.0 - which attempts to import `_pycbf.so` as a local-relative (which it now is) rather than a global module. I've done that in the build step; we probably want to do this in the main repo.
- The only test run is one of importing pycbf - it's hard to know what to run, because the test suite fails on the main repo normally see e.g. https://github.com/yayahjb/cbflib/issues/17, and the CMake test mapping was somewhat out-of-date.
- `tiff2cbf` isn't built (`USE_TIFF=no`). CBFlib uses a custom fork of libtiff that has a unique `TIFFSNPrintDirectory` function, which of course the conda-forge libtiff doesn't have
- Fortran support works locally, but was getting some errors on the feedstock build so temporarily disabled
- Windows build isn't working yet. There seems to be several problems with building this code with `cl.exe`:
    - Apparently doesn't support `#ifdef`s embedded inside macro expansions
    - Dividing by zero is a compiler error
    - Variable length arrays are not supported. This was mandatory in C99, but apparently made optional in C11 so it seems unlikely to be changed (cctbx cbflib/adaptbx does not build the parts of cbflib that depend on this)
    - `unistd.h` is missing. This is used in a few places, mainly for the contents of `sys/stat.h` but possible a few other functions.

Maybe I am just not setting flags, but I think the existing windows builds have been in alternate unix-like environments, like MSys2, maybe? In any case, I have a experimental [branch](https://github.com/ndevenish/cbflib/compare/refactor...ndevenish:winchanges) fixing up some of these as a first attempt, but still encounter linking errors from a couple of non-standard functions that need to be worked around. (This branch also uses [`_alloca`](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/alloca) to, somewhat dangerously, allocate from the stack rather than refactor the functions to properly free heap memory).

Also, the licence is set to dual GPL/LGPL in the existing branch in this repo. I'm not sure how cbflib distinguishes the "API" (LGPL) from "CBFLIB" (GPL), so this might be full GPL2?

ndevenish committed:
Build scripts from fork

ndevenish committed:
Linting

ndevenish committed:
Remove fortran

Anthchirp commented on recipes/cbflib/bld.bat:
```diff
@@ -0,0 +1,19 @@
+
+cd pycbf
+swig -python pycbf.i
```
```suggestion
swig -python pycbf.i || exit /b 1
```

Anthchirp commented on recipes/cbflib/meta.yaml:
```diff
@@ -8,25 +8,40 @@ package:
-    - makecbf --help
+  imports:
+    - pycbf
 
```
```suggestion
  commands:
    - pip check
  requires:
    - pip

```
to ensure environment consistency

ndevenish committed:
Apply suggestions from code review

Co-authored-by: Markus Gerstel <2102431+Anthchirp@users.noreply.github.com>

ndevenish committed:
Use fixbytes branch and fortran for linux

ndevenish commented:
Okay, fixed the fortran build - however, it looks like HDF5 doesn't distribute shared fortran libraries on mac? So I've turned it off except for linux for now. 

I've also moved to point to https://github.com/yayahjb/cbflib/pull/19 - it works off master now, but having the bytes/str fix in makes it in a better state for python3 that setting the swig BYTE flag.

Going to merge this so we can run this on the staged-generator.
